In [1]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from sklearn import linear_model
import statsmodels.formula.api as smf
from sqlalchemy import create_engine

In [9]:
# Display preferences.
%matplotlib inline
pd.options.display.float_format = '{:.3f}'.format
import warnings
warnings.filterwarnings(action="ignore")

postgres_user = 'dsbc_student'
postgres_pw = '7*.8G9QH21'
postgres_host = '142.93.121.174'
postgres_port = '5432'
postgres_db = 'weatherinszeged'

engine = create_engine('postgresql://{}:{}@{}:{}/{}'.format(
    postgres_user, postgres_pw, postgres_host, postgres_port, postgres_db))

temp_df = pd.read_sql_query('select * from weatherinszeged',con=engine)

# no need for an open connection, as we're only doing a single query
engine.dispose()


In [10]:
temp_df.head(50)

date        summary preciptype  temperature  \
0  2006-03-31 22:00:00+00:00  Partly Cloudy       rain        9.472   
1  2006-03-31 23:00:00+00:00  Partly Cloudy       rain        9.356   
2  2006-04-01 00:00:00+00:00  Mostly Cloudy       rain        9.378   
3  2006-04-01 01:00:00+00:00  Partly Cloudy       rain        8.289   
4  2006-04-01 02:00:00+00:00  Mostly Cloudy       rain        8.756   
5  2006-04-01 03:00:00+00:00  Partly Cloudy       rain        9.222   
6  2006-04-01 04:00:00+00:00  Partly Cloudy       rain        7.733   
7  2006-04-01 05:00:00+00:00  Partly Cloudy       rain        8.772   
8  2006-04-01 06:00:00+00:00  Partly Cloudy       rain       10.822   
9  2006-04-01 07:00:00+00:00  Partly Cloudy       rain       13.772   
10 2006-04-01 08:00:00+00:00  Partly Cloudy       rain       16.017   
11 2006-04-01 09:00:00+00:00  Partly Cloudy       rain       17.144   
12 2006-04-01 10:00:00+00:00  Partly Cloudy       rain       17.800   
13 2006-04-01 11:00:00+00:00  Partly Cloudy       rain       17.333   
14 2006-04-01 12:00:00+00:00  Partly Cloudy       rain       18.878   
15 2006-04-01 13:00:00+00:00  Partly Cloudy       rain       18.911   
16 2006-04-01 14:00:00+00:00  Partly Cloudy       rain       15.389   
17 2006-04-01 15:00:00+00:00  Mostly Cloudy       rain       15.550   
18 2006-04-01 16:00:00+00:00  Mostly Cloudy       rain       14.256   
19 2006-04-01 17:00:00+00:00  Mostly Cloudy       rain       13.144   
20 2006-04-01 18:00:00+00:00  Mostly Cloudy       rain       11.550   
21 2006-04-01 19:00:00+00:00  Mostly Cloudy       rain       11.183   
22 2006-04-01 20:00:00+00:00  Partly Cloudy       rain       10.117   
23 2006-04-01 21:00:00+00:00  Mostly Cloudy       rain       10.200   
24 2006-04-09 22:00:00+00:00  Partly Cloudy       rain       10.422   
25 2006-04-09 23:00:00+00:00  Partly Cloudy       rain        9.911   
26 2006-04-10 00:00:00+00:00  Mostly Cloudy       rain       11.183   
27 2006-04-10 01:00:00+00:00  Partly Cloudy       rain        7.156   
28 2006-04-10 02:00:00+00:00  Partly Cloudy       rain        6.111   
29 2006-04-10 03:00:00+00:00  Partly Cloudy       rain        6.789   
30 2006-04-10 04:00:00+00:00  Mostly Cloudy       rain        7.261   
31 2006-04-10 05:00:00+00:00  Mostly Cloudy       rain        7.800   
32 2006-04-10 06:00:00+00:00  Mostly Cloudy       rain        9.872   
33 2006-04-10 07:00:00+00:00  Mostly Cloudy       rain       12.222   
34 2006-04-10 08:00:00+00:00  Mostly Cloudy       rain       15.094   
35 2006-04-10 09:00:00+00:00  Mostly Cloudy       rain       17.356   
36 2006-04-10 10:00:00+00:00  Mostly Cloudy       rain       19.006   
37 2006-04-10 11:00:00+00:00  Mostly Cloudy       rain       20.044   
38 2006-04-10 12:00:00+00:00  Mostly Cloudy       rain       21.050   
39 2006-04-10 13:00:00+00:00  Mostly Cloudy       rain       21.183   
40 2006-04-10 14:00:00+00:00  Mostly Cloudy       rain       20.117   
41 2006-04-10 15:00:00+00:00  Mostly Cloudy       rain       20.217   
42 2006-04-10 16:00:00+00:00  Mostly Cloudy       rain       20.000   
43 2006-04-10 17:00:00+00:00  Mostly Cloudy       rain       17.800   
44 2006-04-10 18:00:00+00:00  Mostly Cloudy       rain       16.061   
45 2006-04-10 19:00:00+00:00  Mostly Cloudy       rain       15.022   
46 2006-04-10 20:00:00+00:00       Overcast       rain       14.422   
47 2006-04-10 21:00:00+00:00       Overcast       rain       14.256   
48 2006-04-10 22:00:00+00:00       Overcast       rain       13.772   
49 2006-04-10 23:00:00+00:00       Overcast       rain       13.283   

    apparenttemperature  humidity  windspeed  windbearing  visibility  \
0                 7.389     0.890     14.120      251.000      15.826   
1                 7.228     0.860     14.265      259.000      15.826   
2                 9.378     0.890      3.928      204.000      14.957   
3                 5.944     0.830     14.104      269.000      15.826   
4                 6.978     0.830    

In [3]:
temp_df.describe()

temperature  apparenttemperature  humidity  windspeed  windbearing  \
count    96453.000            96453.000 96453.000  96453.000    96453.000   
mean        11.933               10.855     0.735     10.811      187.509   
std          9.552               10.697     0.195      6.914      107.383   
min        -21.822              -27.717     0.000      0.000        0.000   
25%          4.689                2.311     0.600      5.828      116.000   
50%         12.000               12.000     0.780      9.966      180.000   
75%         18.839               18.839     0.890     14.136      290.000   
max         39.906               39.344     1.000     63.853      359.000   

       visibility  loudcover  pressure  
count   96453.000  96453.000 96453.000  
mean       10.347      0.000  1003.236  
std         4.192      0.000   116.970  
min         0.000      0.000     0.000  
25%         8.340      0.000  1011.900  
50%        10.046      0.000  1016.450  
75%        14.812      0.000  1021.090  
max        16.100      0.000  1046.380

In [4]:
temp_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 96453 entries, 0 to 96452
Data columns (total 12 columns):
date                   96453 non-null datetime64[ns, UTC]
summary                96453 non-null object
preciptype             96453 non-null object
temperature            96453 non-null float64
apparenttemperature    96453 non-null float64
humidity               96453 non-null float64
windspeed              96453 non-null float64
windbearing            96453 non-null float64
visibility             96453 non-null float64
loudcover              96453 non-null float64
pressure               96453 non-null float64
dailysummary           96453 non-null object
dtypes: datetime64[ns, UTC](1), float64(8), object(3)
memory usage: 7.7+ MB


In [5]:
# 3 objects (categories) and 8 float64 (continous) data types

In [13]:
# missing data
total_missing = temp_df.isnull().sum().sort_values(ascending=False)
percent_missing = (temp_df.isnull().sum()/temp_df.isnull().count()).sort_values(ascending=False)
missing_data = pd.concat([total_missing, percent_missing], axis=1, keys=['Total', 'Percent'])
missing_data.head(13)


Total  Percent
dailysummary             0    0.000
pressure                 0    0.000
loudcover                0    0.000
visibility               0    0.000
windbearing              0    0.000
windspeed                0    0.000
humidity                 0    0.000
apparenttemperature      0    0.000
temperature              0    0.000
preciptype               0    0.000
summary                  0    0.000
date                     0    0.000

In [15]:
# Exploring the data
# Target variable is the difference between the apparenttemperature and the temperature
Y = temp_df['apparenttemperature'] - temp_df['temperature']
# Features: humidity and windspeed
X = temp_df[['humidity', 'windspeed']]

In [18]:
import statsmodels.api as sm
# add constant
X = sm.add_constant(X)

# fit the OLS model using statsmodel
results = sm.OLS(Y,X).fit()

#print summary

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Fri, 20 Dec 2019   Prob (F-statistic):               0.00
Time:                        10:01:53   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.4381      0.021    115.948      0.0

#### the bias is 2.44.
#### Both variables ( humidity & windspeed) are statistically significant as the P value is less than 0.05.
#### Model shows that increase in humidity negatively related to the target.
#### 1 point increase in humidity results in 3.03 decrease in the target.
#### Model also shows that increase in windspeed negatively related to the target.
#### 1 point increase in windspeed results in 0.12 decrease in the target.

In [21]:
# include the interaction of humidity & windspeed to the model
temp_df["humidity_windspeed"]= temp_df.humidity * temp_df.windspeed

# target
Y = temp_df['apparenttemperature'] - temp_df['temperature']

# features
X = temp_df[["humidity_windspeed", "humidity", "windspeed"]]

# add constant
X = sm.add_constant(X)

# fit the OLS model using statsmodel
results = sm.OLS(Y,X).fit()

#print summary

print(results.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Fri, 20 Dec 2019   Prob (F-statistic):               0.00
Time:                        10:20:36   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                  0.0839      0

#### Conclusion
#### Bias have decrease from 2.44 to 0.1
##### As previous model, all variables ( humidity_windspeed, humidity & windspeed) are statistically significant as the P value is less than 0.05
#### For every point drop in humidity_windspeed variable, the target value decrease by 0.3
#### For every point increase in humidity, the target value inreases by 0.12
#### For every point increase in windspeed, the target value increases by 0.1